File: app/src/main/java/com/example/myapplication/data/remote/aidl/UserParcelable.java
package com.example.myapplication.data.remote.aidl;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.Objects;

public class UserParcelable implements Parcelable {

    public static final Creator<UserParcelable> CREATOR = new Creator<UserParcelable>() {
        @Override
        public UserParcelable createFromParcel(Parcel source) {
            return new UserParcelable(source);
        }

        @Override
        public UserParcelable[] newArray(int size) {
            return new UserParcelable[size];
        }
    };

    private final long id;
    private final String name;
    private final int age;
    private final float weight;

    public UserParcelable(long id, String name, int age, float weight) {
        this.id = id;
        this.name = name == null ? "" : name;
        this.age = age;
        this.weight = weight;
    }

    private UserParcelable(Parcel parcel) {
        this.id = parcel.readLong();
        String parcelName = parcel.readString();
        this.name = parcelName == null ? "" : parcelName;
        this.age = parcel.readInt();
        this.weight = parcel.readFloat();
    }

    public long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public int getAge() {
        return age;
    }

    public float getWeight() {
        return weight;
    }

    public UserParcelable copy() {
        return new UserParcelable(id, name, age, weight);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(id);
        dest.writeString(name);
        dest.writeInt(age);
        dest.writeFloat(weight);
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof UserParcelable)) {
            return false;
        }
        UserParcelable that = (UserParcelable) other;
        return id == that.id
                && age == that.age
                && Float.compare(that.weight, weight) == 0
                && Objects.equals(name, that.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, name, age, weight);
    }

    @Override
    public String toString() {
        return "UserParcelable{"
                + "id=" + id
                + ", name='" + name + '\''
                + ", age=" + age
                + ", weight=" + weight
                + '}';
    }
}

File: app/src/main/java/com/example/myapplication/data/remote/aidl/UserParcelable.kt
package com.example.myapplication.data.remote.aidl

import android.os.Parcel
import android.os.Parcelable

data class UserParcelable(
    val id: Long,
    val name: String,
    val age: Int,
    val weight: Float,
) : Parcelable {

    private constructor(parcel: Parcel) : this(
        id = parcel.readLong(),
        name = parcel.readString().orEmpty(),
        age = parcel.readInt(),
        weight = parcel.readFloat(),
    )

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeLong(id)
        parcel.writeString(name)
        parcel.writeInt(age)
        parcel.writeFloat(weight)
    }

    override fun describeContents(): Int = 0

    companion object CREATOR : Parcelable.Creator<UserParcelable> {
        override fun createFromParcel(parcel: Parcel): UserParcelable = UserParcelable(parcel)

        override fun newArray(size: Int): Array<UserParcelable?> = arrayOfNulls(size)
    }
}
